doc/ni_by_example_2.md

```diff
@@ -0,0 +1,395 @@
+#`ni` by Example, Part 2 (WIP)
+
+Welcome to the second part of the tutorial. At this point, you should be familiar with fundamental row and column operations; sorting; basic Perl operations; file I/O and comopression; and some basic principles of `ni` style. Before continuing, it's advisable to read over the horribly misnamed [cheatsheet](cheatsheet.md) to see some of the operations that were glossed over.
+
+The key concept that we will cover (and really, the key to `ni`'s power) is the ability of `ni` to package itself and execute in-memory on remote machines. To that end, we will explain the use of `ni` on local Docker instances; over `ssh` on remote machines, and how to use `ni` to write simple and powerful Hadoop Streaming jobs. 
+
+Other key concepts for this tutorial include streaming reduce operations, data closures, and cell operations. We'll also cover more `ni`-specific Perl extensions, and some important parts of Perl that will be particularly useful in `ni`.
+
+
+To start this tutorial, At this point, it's important to have a general idea of how `ni` works
+
+##`ni` is a quine
+
+A quine is a program that prints its source code. 
+
+
+##`ni` Philosophy and Style
+
+####`ni` demands expertise
+
+####Conciseness matters; readability to the uninitiated does not.
+`ni` spells should be beautiful
+
+####Outsource hard jobs to more appropriate tools.
+The most obvious 
+
+Some jobs that are difficult for `ni`:
+* Sorting
+* Matrix Multiplication
+* SQL Joins
+
+
+
+
+##Perl for `ni` fundamentals
+
+
+
+##Connecting `bash` and `ni`
+
+* `e[<script>]`: Evaluate script
+  * evaluate `<script>` in bash, and stream out the results one line at a time.
+
+##Input Operations
+* `D:<field1>,:<field2>...`: JSON Destructure
+  * `ni` implements a very fast JSON parser that is great at pulling out string and numeral fields.
+  * As of 2016-12-24, the JSON destructurer does not support list-based fields in JSON.
+
+####`ni` and directories
+
+Enrichment sections explore relevant but non-critical `ni` functions. Their contents are not critical to productive `ni` development, and they can be skipped or skimmed.
+
+Start by making some data (Note that you have to be in `bash` for the echo statements to work. [Here](http://stackoverflow.com/questions/8467424/echo-newline-in-bash-prints-literal-n) is a very interesting post about `echo`'s unintuitive behavior):
+
+```
+$ rm -rf dir_test && mkdir dir_test
+$ echo -e "hello\n" > dir_test/hi
```

Comment:
`echo` normally appends a newline; e.g. `echo foo > bar` results in `bar` containing the four bytes `foo\n`.